### PR TITLE
Implement `uv build`

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This PR exposes uv's PEP 517 implementation via a `uv build` frontend, such that you can use `uv build` to build source and binary distributions (i.e., wheels and sdists) from a given directory.

There are some TODOs that I'll tackle in separate PRs:

- [x] Support building a wheel from a source distribution (rather than from source) (#6898)
- [x] Stream the build output (#6912)

Closes https://github.com/astral-sh/uv/issues/1510

Closes https://github.com/astral-sh/uv/issues/1663.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Needs tests before it's ready for review, but the behavior is there.

**Comment by charliermarsh** [comment]
I will do a pass on the docs once the interface gets the green light.

File: crates/uv/src/commands/build.rs
Comment (charliermarsh) [inline crates/uv/src/commands/build.rs:245]: Do we want to show "Building wheel"? I think so -- should it include a progress spinner? (Also think so.)

Do we want to show the build _output_ by default? (As of #6903 it will show in `--verbose`.)

**Comment by charliermarsh** [comment]
I plan to show build output in a later PR.

File: crates/uv/tests/build.rs
Comment (konstin) [inline crates/uv/tests/build.rs:2]: This seems to have no effect on my machine

File: crates/uv-cli/src/lib.rs
Comment (konstin) [inline crates/uv-cli/src/lib.rs:342]: Where can I find these files after they are built?

File: crates/uv-cli/src/lib.rs
Comment (konstin) [inline crates/uv-cli/src/lib.rs:2935]: The whole block seems to be dead code?

File: crates/uv/src/commands/build.rs
Comment (konstin) [inline crates/uv/src/commands/build.rs:60]: Should we should the relative path here instead of the filename?

File: crates/uv/src/commands/build.rs
Comment (konstin) [inline crates/uv/src/commands/build.rs:238]: We should explain this behavior in the cli docs

File: crates/uv-client/src/base_client.rs
Comment (konstin) [inline crates/uv-client/src/base_client.rs:194]: That's from a different PR

File: crates/uv/src/commands/build.rs
Comment (konstin) [inline crates/uv/src/commands/build.rs:230]: We should make this configurable and add a `.gitignore` by default (similar to cache dirs)

**Review by konstin — APPROVED** [review]

**Comment by konstin** [comment]
Should `uv build` build the entire workspace or only the current package?

File: crates/uv/src/commands/build.rs
Comment (charliermarsh) [reply crates/uv/src/commands/build.rs:230]: Will do RE `outdir`. Not sure on `.gitignore `though...

**Comment by zanieb** [comment]
I would imagine building the entire workspace by default and `--package` to select a specific one?

**Comment by charliermarsh** [comment]
I’d prefer to tackle any workspace integration in a separate PR. How would that intersect with the ability to pass a single source dist file? Or the ability to pass a path to a directory?

File: crates/uv-cli/src/lib.rs
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:1944]: Should we just say "distributions"
```suggestion
    /// The directory from which distributions should be built.
```

File: crates/uv/src/commands/build.rs
Comment (zanieb) [inline crates/uv/src/commands/build.rs:128]: Can we support `--no-project` here? Trying to do that everywhere. Might not make sense for this command though?

File: crates/uv/src/commands/build.rs
Comment (zanieb) [inline crates/uv/src/commands/build.rs:129]: Should this be searching in the target directory instead of the working directory?

**Comment by zanieb** [comment]
Another todo is to add documentation to

- The publishing guide
- The project concept

**Comment by zanieb** [comment]
> How would that intersect with the ability to pass a single source dist file? Or the ability to pass a path to a directory?

I think if we're give a file target, we can just build that. If we're given a directory, we should treat that as the working directory for workspace / project discovery?

**Comment by charliermarsh** [comment]
I will add `--outdir` separately.

**Comment by charliermarsh** [comment]
Nevermind, I added `--out-dir` support here. I will do workspaces in another PR.

**Comment by charliermarsh** [comment]
Ok, I added docs in https://github.com/astral-sh/uv/pull/6991, and workspace support (just `--package` for now) in #6990, so in theory the whole stack is ready to merge and ship once approved.

File: crates/uv/src/commands/build.rs
Comment (zanieb) [inline crates/uv/src/commands/build.rs:186]: Fyi there is a `impl<'a> TryFrom<BaseClientBuilder<'a>> for RegistryClientBuilder<'a>` so we can avoid some repetition here? Kind of annoying it creates a temporary cache though.

**Review by zanieb — APPROVED** [review]

**Comment by charliermarsh** [comment]
Again `PermissionError: [Errno 13] Permission denied: 'E:\/uv-tmp\/[TMP]/__init__.py'`.

**Pull request merged**
